pycalphad/mapping/zpf_checks.py

```diff
@@ -0,0 +1,466 @@
+    """
+    Check if the point is global minimum
+    1. Check if a new composition set can be added that can lower free energy
+    2. Create a new node with the additional composition set
+    3. Check that the new node is valid
+
+    3 possible outcomes
+        a) No change in phases -> pass
+        b) Change in phases, node successfully found -> process new node and end zpf line with graceful ending
+        c) Change in phases, node not found -> end zpf line with unexpected ending
+
+    Parameters
+    ----------
+    zpf_line : ZPFLine
+        ZPFLine that the point is stepping in
+    step_results : [Point, [CompositionSet]]
+        Results from zpf_equilibrium.update_equilibrium_with_new_conditions
+    axis_data : dict
+        Axis variable data from a map strategy class
+
+    Returns
+    -------
+    new_node : Node
+        Node from step_results if global min is found
+    
+    or
+
+    None : if step_results is still global min
+           or new node cannot be found (zpf_line will be ended)
+    """
+    if step_results is None:
+        return None
+
+    axis_vars, axis_delta, axis_lims = axis_data["axis_vars"], axis_data["axis_delta"], axis_data["axis_lims"]
+    global_check_interval = kwargs.get("global_check_interval", 1)
+    pdens = kwargs.get("pdens", 500)
+    tol = kwargs.get("tol", 1e-4)
+    num_candidates = kwargs.get("global_num_candidates", 1)
+    system_info = kwargs.get("system_info", None)
+    do_not_create_node = kwargs.get("do_not_create_node", False)    #For testing/debugging purposes
+
+    new_point, orig_cs = step_results
+    new_point_vars = {av: new_point.get_property(av) for av in axis_vars}
+    if len(zpf_line.points) % global_check_interval == 0:
+        global_test_point = zeq.find_global_min_point(new_point, system_info, pdens, tol, num_candidates)
+        if global_test_point is not None:
+            _log.info(f"Global min detected. {new_point.stable_phases} -> {global_test_point.stable_phases}")
+            # By default, assumed the zpf failed. When we successfully find a new node, then we change this
+            zpf_line.status = ZPFState.FAILED
+
+            if do_not_create_node:
+                new_node = None
+            else:
+                new_node = zeq.create_node_from_different_points(global_test_point, new_point.stable_composition_sets, axis_vars)
+
+            if new_node is not None:
+                new_node_vars = {av: new_node.get_property(av) for av in axis_vars}
+                # Check that new node satisfy axis limits and distance between nodes
+                # If not, then the zpf line ends unexpectedly
+                check_axis = _check_axis_values_within_limit(zpf_line, new_point_vars, new_node_vars, axis_data, **kwargs)
+                check_dist =_check_axis_values_by_distance(zpf_line, new_point_vars, new_node_vars, axis_data, **kwargs)
```

Comment:
Can we check against previously removed nodes somehow?

Reply:
I suppose rather than removing them, we can just flag them as bad nodes (which will be ignored during plotting) in the NodeQueue, so the NodeQueue will check against all existing nodes when a new node is attempted to be added. The only thing for binary cases is that we would need a way to determine that the node is metastable and flag them.